### DISCUSSION TITLE
What's the current state of vim keybindings in Theia?

### DISCUSSION THREAD

**Discussion by peey** [eclipse-theia/theia#7829] [General]
Here's a summary of what I know:

VIM VSCode extension works in theia (https://github.com/eclipse-theia/theia/issues/5125) but has damned performance (https://github.com/eclipse-theia/theia/issues/5125#issuecomment-595450866, https://github.com/gitpod-io/gitpod/issues/1212) due to difference in architecture (the extension sends every keystroke to the server and waits for the response  (https://github.com/gitpod-io/gitpod/issues/1212#issuecomment-585631411) whereas native editing in theia doesn't).

I found some old issues discussing editor-level support for vim keybindings (https://github.com/eclipse/che/issues/9244) and the approach taken would be to implement a plugin ([che-theia-plugin](https://github.com/eclipse/che-theia-vi-plugin)) and add a Keyboard Shortcuts API (https://github.com/eclipse-theia/theia/issues/2783). Unfortunately, the planned plugin is not in active development (https://github.com/eclipse/che-theia-vi-plugin/pull/1#issuecomment-425014412).

Is there any active effort to develop a vim keybindings plugin?

**Comment by akosyakov** [comment]
@peey I think it is the pretty much the current state. There are several ways to move forward:
- wait till VS Code team figures out how to resolve it and then port: https://github.com/microsoft/vscode/issues/75627
- try to resolve via a Theia extension which runs VIM VS Code extension not in the same process as other VS Code extensions, maybe even no on the backend but in dedicated web worker

Most likely we will wait till the issue is sorted out in the upstream.

**Reply by peey** [reply]
That's a sensible choice for the  VSCode VIM extension. 

From a quick glance, it appears that there are also "Theia Extensions" which don't share weaknesses of VSCode's extension architecture. I don't know about these in detail, could you tell me if another way to move forward would be to build a Theia Extension for vim emulation?

**Reply by akosyakov** [reply]
Yes, it would be another option to create a Theia extension which runs VIM logic in web worker. The thing is that Theia extension is intended to provide custom Theia based product, so they are built in the product and cannot be removed. VS Code extensions can be installed at runtime. They are not so powerful and has to be secured but they can be shared between Theia based products. So if you go for Theia extension, you should make sure that it is can be enabled/disabled via settings, products can decide about defaults.

Another downside of this approach that there is no community supporting it out of the box, so you will need to build it. Maybe you can reuse code from VS Code VIM extension somehow.

**Reply by peey** [reply]
That makes sense. I just wanted to know what's technically possible. 

I was going to suggest that Theia should have its own extensions system, so that while it supports VSCode extensions, people can also build Theia extensions which maybe don't work with VSCode. 

The reason behind this would be that:

1. There is an important distinction between Theia and VSCode that the former is supposed to run in a browser and latency is incurred when communicating with the server.
2. Theia team has no ownership or control over VSCode extensions and their architecture. 

Majority of the extensions will be unaffected by #1 and work just as well with both Theia and VSCode, but for those that are affected by it, there should be a way to support them in Theia. VIM keybindings support  is one such extension.

But I'm **not** going to suggest it because it doesn't seem like it's completely blocked. I think we can achieve just as much by making a Theia extension and shipping it with an option to be disabled or enabled, just like you mentioned. I hope that's more or less correct.
